packages/react-utilities/src/hooks/useControllableState.ts

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import { useConst } from './useConst';
+
+type NonUndefined<T> = T extends undefined ? never : T;
+
+type UseAutoControlledOptions<ControllableState, State> = {
+  /**
+   * User provided default state or factory initializer
+   */
+  defaultState?: State | (() => State);
+  /**
+   * User provided controllable state, undefined state means internal state will be used
+   */
+  state: ControllableState;
+  /**
+   * Used to initialize state if all user provided states are undefined
+   */
+  initialState: State;
+};
+
+function isUndefined(state: unknown): state is undefined {
+  return typeof state === 'undefined';
+}
+
+function isFactoryDispatch<State>(newState: React.SetStateAction<State>): newState is (prevState: State) => State {
+  return typeof newState === 'function';
+}
+
+/**
+ * A useState 'like' hook that allows optional user control
+ * Useful for components which allow uncontrolled and controlled behaviours for users
+ * @returns - https://reactjs.org/docs/hooks-state.html
+ */
+export const useControllableState = <ControllableState, State extends NonUndefined<ControllableState>>(
+  options: UseAutoControlledOptions<ControllableState, State>,
```

Comment:
```suggestion
  options: useControllableStateOptions<ControllableState, State>,
```

Oops, a type name does not match...

Reply:
Can we update generics there to accept a single param as in `React.useState()`? Is there any reasoning to have both?

Currently it's unclear what types should be passed:

```tsx
const [open, setOpen] = useControllableState<boolean, /* 💭 what should be there? */>()
```

Reply:
> Oops, a type name does not match... 

can you elaborate pls?

Reply:
> Can we update generics there to accept a single param as in `React.useState()`? Is there any reasoning to have both?
> 
> Currently it's unclear what types should be passed:
> 
> ```tsx
> const [open, setOpen] = useControllableState<boolean, /* 💭 what should be there? */>()
> ```

`useState` uses function definition override for `undefined`, we need 2nd generic to get rid of that `undefined` from state.

What I'd suggest for this API is to provide default generic value, so users can still have 1 generic API

```ts
export const useControllableState = <
  ControllableState,
  State extends NonUndefined<ControllableState> = NonUndefined<ControllableState>
>(
```

Reply:
The whole point of `initialState` being a required prop is that inference would start working, but using a default generic does make sense.

@layershifter  northstar did not need the second generic type because it does not use strict null checks.

Reply:
Kudos to @layershifter  for using a union with `undefined` to avoid the second generic 🚀🚀